### PR TITLE
SHM refine: Allocation API

### PR DESCRIPTION
**Pull request by yellowhatter**

### CODE REVIEW

**Comment by yellowhatter** [comment]
@Mallets this is a new shm alloc API

File: examples/examples/z_alloc_shm.rs [hunk 1]
```diff
@@ -34,20 +34,51 @@ async fn run() -> ZResult<()> {
...
-        // OPTION 2: Comprehensive configuration:
+        // OPTION: Comprehensive configuration:
         let _comprehensive_layout = provider
-            .alloc_layout()
-            .size(512)
-            .alignment(AllocAlignment::new(2))
-            .res()
+            .alloc(512)
+            .with_alignment(AllocAlignment::new(2))
+            .make_layout()
```
Comment (Mallets) [inline examples/examples/z_alloc_shm.rs:81]: What about calling it `into_layout`?

File: commons/zenoh-shm/src/api/provider/shared_memory_provider.rs [hunk 2]
```diff
@@ -64,83 +67,69 @@ impl BusyChunk {
...
+    pub fn with_policy<Policy>(self) -> AllocBuilder2<'a, IDSource, Backend, Policy> {
+        AllocBuilder2 {
+            data: self.0,
+            _phantom: PhantomData,
+        }
+    }
+
+    /// Get the result
+    #[zenoh_macros::unstable_doc]
+    pub fn res(self) -> BufLayoutAllocResult {
```
Comment (Mallets) [inline commons/zenoh-shm/src/api/provider/shared_memory_provider.rs:126]: https://github.com/eclipse-zenoh/zenoh/pull/942 has been recently merged. I believe we should use either `.wait()` or `.await` from now on.

File: commons/zenoh-shm/src/api/provider/shared_memory_provider.rs [hunk 3]
```diff
@@ -511,6 +501,75 @@ unsafe impl<'a, Policy: AllocPolicy, IDSource, Backend: SharedMemoryProviderBack
...
+// Async Alloc policy
+impl<'a, IDSource, Backend, Policy> AllocBuilder2<'a, IDSource, Backend, Policy>
+where
+    IDSource: ProtocolIDSource + Send + Sync,
+    Backend: SharedMemoryProviderBackend + Sync,
+    Policy: AsyncAllocPolicy,
+{
+    /// Get the async result
+    #[zenoh_macros::unstable_doc]
+    pub async fn res_async(self) -> BufLayoutAllocResult {
```
Comment (Mallets) [inline commons/zenoh-shm/src/api/provider/shared_memory_provider.rs:561]: Same comment as above. We should use either `.await` or `.wait()` by implementing `Wait` or `IntoFuture` traits.

File: commons/zenoh-shm/src/api/provider/shared_memory_provider.rs (hunk 2 above)
Comment (yellowhatter) [reply commons/zenoh-shm/src/api/provider/shared_memory_provider.rs:126]: I know about our new `Wait` and `IntoFuture` traits, I was just concerned that `.wait()` will confuse our users that it blocks when it doesn't.

Practically it blocks only if BlockOn policy is used, so we have three types of behaviors:
- non-blocking (any policy except BlockOn)
- blocking (only BlockOn policy):
- - sync
- - async (available for `Backend: Sync`)

I think I need to split allocation API into these three forms:
- `res()` for non-blocking policies
- `Wait` for sync BlockOn
- `IntoFuture` for async BlockOn

Comment (Mallets) [reply commons/zenoh-shm/src/api/provider/shared_memory_provider.rs:126]: Regardless it's blocking or non-blocking, `.wait()` is the common method for resolving builders in `sync` context in the same way you call `.await` in `async` context. I don't see much benefit for adding another way to resolve the builder...

**Comment by yellowhatter** [comment]
macos-13 test failure is a regression that is fixed in main

**Pull request merged**